### PR TITLE
Add support for package.json #200

### LINKED ISSUE #200: Add support for package.json
In order to reduce the number of false positives when scanning images while also keeping the value of grype-vscode [directory] scans, it would be useful when scanning directories to key off of index-like files (e.g. `package-lock.json`) and installation metadata (e.g. `package.json` within `node_module` sub-directories) when scanning images for the javascript cataloger.

**AC**
- Dependencies described within `yarn.lock` and `package-lock.json` files are not shown during image scans (but are shown during directory scans).
- The package described in `package.json` is listed in the SBOM for both image and directory scans.
- NO dependencies listed in the `package.json` should be present in the SBOM.

**Steps to Test**
- When scanning a directory:
  - ensure there are examples of a `package.json`, `yarn.lock`, `package-lock.json` files in the directory being scanned.
  - ensure the scan finds all javascript example packages.
- When scanning an image: 
  - ensure there are examples of a `package.json`, `yarn.lock`, `package-lock.json` files in the image being scanned.
  - ensure the scan finds all javascript package examples EXCEPT any `package-lock.json` and `yarn.lock` references.

**Notes for Developer**
Should mirror the gemspec cataloger additions made in https://github.com/anchore/syft/pull/203 

Today we capture npm packages from discovered `package-lock.json` and `yarn.lock` files (which includes top-level dependencies and their dependencies). We should additionally use `package.json` to discover top-level dependencies, tracking only the package the `package.json` represents and ignoring any listed dependencies.

### PR DESCRIPTION
**Pull request by Toure**
This patch-set adds support for package.json, it also creates a separation between directory and image-based information, which allows for the removal of yarn pkg type and replaces it with npm as a new hierarchy. The last change made was the removal of the Golang build exclusion tag " +build integration" this was possible through the use of a _test suffixes on test packages.

Signed-off-by: Toure Dunnon <toure.dunnon@anchore.com>
Closes #200

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -130,7 +130,7 @@ unit: fixtures ## Run unit tests (with coverage)
 .PHONY: integration
 integration: ## Run integration tests
 	$(call title,Running integration tests)
-	go test -tags=integration ./test/integration
```
Comment (alfredodeza) [inline Makefile:133]: what happened here?

File: schema/json/schema.json [hunk 2]
```diff
@@ -40,6 +40,9 @@
                             "architecture": {
                                 "type": "string"
                             },
+                            "author": {
```
Comment (alfredodeza) [inline schema/json/schema.json:43]: this means it accepts both "author" and "authors" ?

File: syft/cataloger/javascript/parse_package_json.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+		if err := dec.Decode(&p); err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, fmt.Errorf("failed to parse package.json file: %w", err)
+		}
+
+		packages = append(packages, pkg.Package{
+			Name:    p.Name,
+			Version: p.Version,
+			//Source: TODO?
```
Comment (wagoodman) [inline syft/cataloger/javascript/parse_package_json.go:43]: nit: unnecessary comment? (or unclear)

File: syft/cataloger/javascript/parse_package_json.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	License      string            `json:"license"`
+	Name         string            `json:"name"`
+	Homepage     string            `json:"homepage"`
+	Description  string            `json:"description"`
+	Dependencies map[string]string `json:"dependencies"`
+}
+
+// parsePackageJson parses a package.json and returns the discovered JavaScript packages.
+func parsePackageJSON(_ string, reader io.Reader) ([]pkg.Package, error) {
+	packages := make([]pkg.Package, 0) // make empty slice of type pkg.Package
```
Comment (wagoodman) [inline syft/cataloger/javascript/parse_package_json.go:29]: nit: these couple of comments for the slice and reader should probably be removed

File: syft/cataloger/javascript/cataloger.go [hunk 5]
```diff
@@ -8,11 +8,20 @@ import (
...
+
+// NewJavascriptCataloger returns a new Javascript cataloger based on package.json.
+func NewJavascriptLockCataloger() *common.GenericCataloger {
+	globalParsers := map[string]common.ParserFn{
 		"**/package-lock.json": parsePackageLock,
 		"**/yarn.lock":         parseYarnLock,
 	}
 
-	return common.NewGenericCataloger(nil, globParsers, "javascript-cataloger")
+	return common.NewGenericCataloger(nil, globalParsers, "javascript-lock-cataloger")
```
Comment (alfredodeza) [inline syft/cataloger/javascript/cataloger.go:26]: I am not entirely sure that `globalParsers` is accurate here, did you mean to use `globParsers` instead? This is coming from the usage of globs...

File: syft/pkg/type.go [hunk 6]
```diff
@@ -12,11 +12,11 @@ const (
...
-	RpmPkg                Type = "rpm"
-	WheelPkg              Type = "wheel"
-	PoetryPkg             Type = "poetry"
-	NpmPkg                Type = "npm"
-	YarnPkg               Type = "yarn"
+	RpmPkg    Type = "rpm"
+	WheelPkg  Type = "wheel"
+	PoetryPkg Type = "poetry"
+	NpmPkg    Type = "npm"
+	//YarnPkg               Type = "yarn"
```
Comment (wagoodman) [inline syft/pkg/type.go:19]: nit: this should be deleted instead of commented out

File: syft/cataloger/javascript/cataloger.go (hunk 5 above)
Comment (wagoodman) [inline syft/cataloger/javascript/cataloger.go:26]: nit: [`glob`](https://en.wikipedia.org/wiki/Glob_(programming)) doesn't stand for `global` in this case (should probably stay as `glob`)

File: schema/json/schema.json (hunk 2 above)
Comment (wagoodman) [reply schema/json/schema.json:43]: yeah, this section is shared with all metadata's... one has author, another has authors. (this looks valid to me)

File: Makefile (hunk 1 above)
Comment (wagoodman) [reply Makefile:133]: essentially by renaming all files in the directory with `_test` we no longer need the `integration` build tag (🎉 ) which has been a common annoyance. I suggested this change on the last pairing session this morning.

File: syft/cataloger/javascript/cataloger.go [hunk 7]
```diff
@@ -8,11 +8,20 @@ import (
...
-func NewJavascriptCataloger() *common.GenericCataloger {
+func NewJavascriptPackageCataloger() *common.GenericCataloger {
 	globParsers := map[string]common.ParserFn{
+		"**/package.json": parsePackageJSON,
+	}
+
+	return common.NewGenericCataloger(nil, globParsers, "javascript-package-cataloger")
+}
+
+// NewJavascriptCataloger returns a new Javascript cataloger based on package.json.
```
Comment (luhring) [inline syft/cataloger/javascript/cataloger.go:19]: This doc comment needs updating, both the name and the description

File: schema/README [hunk 8]
```diff
@@ -0,0 +1,9 @@
+## Updating the JSON schema
```
Comment (wagoodman) [inline schema/README:1]: nit: this should be `README.md`

File: syft/cataloger/javascript/cataloger.go [hunk 9]
```diff
@@ -7,12 +7,21 @@ import (
...
+func NewJavascriptPackageCataloger() *common.GenericCataloger {
+	globParsers := map[string]common.ParserFn{
+		"**/package.json": parsePackageJSON,
+	}
+
+	return common.NewGenericCataloger(nil, globParsers, "javascript-package-cataloger")
+}
+
+// NewJavascriptCataloger returns a new Javascript cataloger object base on package lock files.
+func NewJavascriptLockCataloger() *common.GenericCataloger {
```
Comment (luhring) [inline syft/cataloger/javascript/cataloger.go:20]: The doc comment should match the function name ("NewJavascriptLockCataloger")

File: syft/cataloger/javascript/cataloger.go [hunk 10]
```diff
@@ -7,12 +7,21 @@ import (
 	"github.com/anchore/syft/syft/cataloger/common"
 )
 
-// NewJavascriptCataloger returns a new JavaScript cataloger object.
-func NewJavascriptCataloger() *common.GenericCataloger {
+// NewJavascriptCataloger returns a new JavaScript cataloger object based on detection of npm based packages.
+func NewJavascriptPackageCataloger() *common.GenericCataloger {
```
Comment (luhring) [inline syft/cataloger/javascript/cataloger.go:11]: The doc comment should match the function name ("NewJavascriptPackageCataloger"). Also, we might want to specify in the description that this cataloger looks at `package.json` files.

File: syft/cataloger/javascript/cataloger.go (hunk 9 above)
Comment (Toure) [reply syft/cataloger/javascript/cataloger.go:20]: Thanks, good catch.

File: syft/cataloger/javascript/cataloger.go (hunk 10 above)
Comment (Toure) [reply syft/cataloger/javascript/cataloger.go:11]: Thanks, good catch.

**Review by wagoodman — APPROVED** [review]

**Pull request merged**